### ISSUE TITLE
Multiple accidentals in tonics, expanded support for modes in note spelling

### ISSUE THREAD

**Issue by bmcfee** [librosa/librosa#1724]
**Is your feature request related to a problem? Please describe.**

Our note spelling code currently supports only major and minor keys, and within those, it only supports at most one accidental in the tonic.

For example, the following all work:

```python
In [2]: librosa.key_to_notes("C:maj")
Out[2]: ['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']

In [3]: librosa.key_to_notes("C:min")
Out[3]: ['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']

In [4]: librosa.key_to_notes("Cb:min")
Out[4]: ['C', 'D♭', 'E𝄫', 'E♭', 'F♭', 'F', 'G♭', 'A𝄫', 'A♭', 'B𝄫', 'B♭', 'C♭']

In [5]: librosa.key_to_notes("C#:min")
Out[5]: ['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']
```

but multiple accidentals do not work:

```python
In [6]: librosa.key_to_notes("C##:min")
[...]
ParameterError: Improper key format: C##:min
```

*Why does this matter?*

Major and minor cover a lot of use-cases, but it would be nice if we could automate note spelling in different modes.  This comes up in the specshow examples of the trumpet clip, which is properly in `F:dorian`, but we translate to `Eb:major` for the purpose of chroma visualization:
![image](https://github.com/librosa/librosa/assets/1190540/9d4d64bc-4330-4a81-b167-e111538a4e61)

It would be great if we could just say `key='F:dorian'` and let the package figure out the translation, rather than having to work out the conversion first manually.

On the surface, this seems like it could be easily implemented by a straightforward table look-up: just shift over to whatever the relative major key is and we're done.  Under the hood, this is already how our note spelling for minor keys works, by displacing the tonic by 9 semitones:

https://github.com/librosa/librosa/blob/519d6d97b0dc9bde42445513bcdfdb9b921d8b7f/librosa/core/notation.py#L567-L571

E.g., `D:dorian → C:major`, etc.  The problem arises when we have accidentals in the starting tonic, which can sometimes translate to double-accidentals in the relative major tonic: `Cb:dorian → Bbb:major`, and these are currently not supported by our parser.

**Describe the solution you'd like**

If we can generalize the tonic parser to support multiple accidentals, this should become much easier to handle.  We'd then be able to extend `key_to_notes` to support modes by coding in the number of semitones (and perhaps separately, diatonic steps) needed to move to the relative major.  

`key_to_degrees` could also be extended, though this is much easier as it should only require building out the lookup table we already have here:

https://github.com/librosa/librosa/blob/519d6d97b0dc9bde42445513bcdfdb9b921d8b7f/librosa/core/notation.py#L685-L687

**Describe alternatives you've considered**

At some point, I took a stab at implementing this by using a parallel major conversion with post-hoc corrections (eg: `C:lydian` = `C:maj` + raised fourth), rather than a relative major.  This turned out to not work for 16 out of 147 cases (7 roots, 7 modes, flat|sharp|natural).
I expect that fixing this approach would be much more complicated than the proposed solution above.

**Additional context**
This has been lurking in my stash of prototyping notebooks for quite some time (eg since 2021 at least), and I only today realized that I never wrote up an issue about it.  :shrug:

**Comment by erabinov** [comment]
Hi! This sounds like a task that I can handle in terms of the theoretical aspects. But this is my first time contributing, so I may need a little bit of guidance on that end. I'll aim to prototype something today and share it via gist, but first I have a few questions:

1. Do you want to enable something like`'C#b#'`as a parameter? Or is it enough to just handle double accidentals?
2. Are `key_to_degree` and `key_to_notes` the only two methods you want changed?

**Comment by erabinov** [comment]
Ok, so here's a first pass at solving this issue. I use recursion: for "Cbb:maj", I add flats to all the notes in the return value for "Cb:maj", then cyclically permute the notes so that the enharmonic equivalent of C comes first.

I needed to make a number of simple helper functions, and I'm still a bit unsure about what formatting to allow for each function, so I needed to add a couple of other REs to the notation.py file.

[Here's a gist](https://gist.github.com/erabinov/2291d6943d2b406933779ead0ef9ea86) that shows the relevant additions. It's not meant to stand alone. The most important changes are in lines 80-100.
